modules/ding_config_import/ding_config_import.module

```diff
@@ -0,0 +1,169 @@
+<?php
+
+/**
+ * @file
+ * Implementation of simple form to upload YAML file.
+ */
+
+use Symfony\Component\Yaml\Yaml;
+
+/**
+ * Implements hook_menu().
+ */
+function ding_config_import_menu() {
+  $item = [];
+
+  $item['admin/config/ding/config'] = [
+    'title' => 'Upload configuration',
+    'description' => 'Upload and import configuration from YAML.',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('ding_config_import_admin_form'),
+    'access arguments' => array('administer site configuration'),
+  ];
+
+  return $item;
+}
+
+/**
+ * Administration form for yaml file upload.
+ */
+function ding_config_import_admin_form() {
+  $url = drupal_get_path('module', 'ding_config_import') . '/example.config.yml';
+
+  $form = [
+    '#attributes' => [
+      'enctype' => 'multipart/form-data',
+    ],
+    'wrapper' => [
+      '#type' => 'fieldset',
+      '#title' => t('Configuration upload'),
+      '#description' => t('This module comes with an <a href="/@url">example file</a> that can be used as an template for the upload. Note that enabled/disable modules may timeout, if it takes longer than the web-servers configuration allows.', ['@url' => $url]),
+    ],
+  ];
+
+  $form['wrapper']['upload'] = [
+    '#title' => t('Upload configurations file (yml)'),
+    '#type'  => 'file',
+  ];
+
+  $form['submit_upload'] = array(
+    '#type' => 'submit',
+    '#value' => t('Upload'),
+  );
+
+  return $form;
+}
+
+/**
+ * Form submit handler.
+ */
+function ding_config_import_admin_form_submit($form, &$form_state) {
+  $error = false;
+  $msg = t('The configuration options have been saved. You may need to clear cache to rebuild things if you have enabled/disabled modules etc.');
+  $file = $_FILES['files']['tmp_name']['upload'];
+
+  if (libraries_load('yaml')) {
+    $data = file_get_contents($file);
+    $values = Yaml::parse($data);
+    if (is_array($values)) {
+      _ding_config_import_set_values($values);
+    }
+    else {
+      $error = true;
+      $msg = t('Fail to parse the YAML file.');
+    }
+  }
+  else {
+    $error = true;
+    $msg = t('Fail to load YAML parsing library. Please contact the system administrator.');
+  }
+
+  if ($error) {
+    drupal_set_message($msg, 'error');
+  }
+  else {
+    drupal_set_message($msg, 'status');
+  }
+
+  // Ensure tmp file is removed. It contains secrets.
+  @unlink($file);
+}
+
+/**
+ * Recursive walk the array using the key as prefix.
+ *
+ * If this an multilevel array the keys are joined with "_". If the flatten
+ * parameter is FALSE.
+ *
+ * @param array $values
+ *   The values.
+ * @param string $prefix
+ *   The prefix, array key form previous recursion.
+ */
+function _ding_config_import_set_values(array $values, $prefix = '') {
+  foreach ($values as $key => $value) {
+    switch ($key) {
+      case 'modules':
+        // Disable modules.
+        if (!empty($value['disable'])) {
+          module_disable($value['disable']);
+        }
```

Comment:
I do not think it is a good idea to have module enabling/disabling as a part of the recursive function. It increases the complexity of the function and can lead to problems if a key within an array intended for variables is called `modules`.

I suggest using separate root elements `variables`/`settings` and `modules` and using separate functions for handling each of them. This also allows for further expansion down the line.

Reply:
Updated to have root elements `settings` and `modules`

Reply:
:+1: I think you should update the settings examples in the documentation accordingly.